### ISSUE TITLE
[FEATURE_REQUEST] Missing or inaccurate tooltips

### ISSUE THREAD

**Issue by Technologicat** [SillyTavern/SillyTavern#1717]
### Have you searched for similar requests?

Yes

### Is your feature request related to a problem? If so, please describe.

Some features are not self-explanatory, but are currently missing a tooltip in the GUI. Also, some existing tooltips are placeholders (*"helpful tip coming soon"*) or could be improved.

### Describe the solution you'd like

Improve the tooltips. :)

I can write some of them myself and submit a PR, but ST has many features, and I don't even use the majority of them. Therefore, there are many features that are complete question marks for me, and I'll need input from other users and devs as to what the features are supposed to do.

If someone can provide that context (by replying here), I can try to condense the information into a tooltip.

### Additional context

Here are the ones known so far:

- **MovingUI** (*User Settings ⊳ Advanced*):
  - "Allow repositioning certain UI elements by dragging them."
- **MUI Preset**:
  - "MovingUI preset. Predefined/saved draggable positions."
  - ~Is this a theme selector for MovingUI, affecting how the dragging GUI looks, or something else?~
- **No WI/AN** (Extensions ⊳ Summarize ⊳ Summary settings)
  - "Do not summarize World Info and Author's Note.", I suppose?
- **Depth** (appears in many places):
  - "How many messages before the current end of the chat."
  - I think this is important to clarify, because at least to a programmer, "depth" first brings to mind nested brackets; and brackets are actually used in ST, to make parenthetical remarks to the AI (such as for summarization: *"[Pause your roleplay. Write a summary...]*").
- **AI Response Configuration**:
  - **Top P**: Otherwise the current tooltip is fine, but maybe mention that Top P is also known as *nucleus sampling*.
  - **Top A**: Relative of **Min P**, but operates on squared probabilities.
    - See e.g. https://www.reddit.com/r/KoboldAI/comments/vcgsu1/comment/icrp0n1
  - **Tail Free Sampling**:
    - "Estimates where the 'knee' of the next-token probability distribution is, and cuts the tail off at that point."
    - I would assume the slider controls the `z` value, but this should be confirmed from the source code.
    - See https://www.trentonbricken.com/Tail-Free-Sampling/
  - **Typical P** = ???
  - **Epsilon Cutoff** = ???
  - **Eta Cutoff** = ???
  - **Mirostat**:
    - "Thermostat for output perplexity. Controls the output perplexity directly, to match the perplexity of the input, thus avoiding the repetition trap (where, as the autoregressive inference produces text, the perplexity of the output tends toward zero) and the confusion trap (where the perplexity diverges)."
    - See https://arxiv.org/abs/2007.14966
    - In practice, **Min P** can lead to similarly good results, while being simpler and faster. Should we mention this?
  - **Beam Search** = ???
    - At least it's the name of a classical optimization method in numerics. Also, in LLM sampling, beam search is infamous for its bad performance; Beam Search easily gets stuck in a repetition loop (which hints that it always picks tokens that are too probable, decreasing output perplexity). I think this was mentioned in one of the Contrastive Search papers (see below).
  - **Contrast Search**: "The representation space of most LLMs is isotropic, and this sampler exploits that in order to encourage diversity while maintaining coherence."
    - Name should be "**Contrastive** Search"
    - In math terms, this is a minor modification to an older, standard sampling strategy. I'll have to re-read the paper to check details. What I do recall is that the penalty alpha controls the relative strength of the regularization term.
    - See https://arxiv.org/abs/2202.06417 , https://arxiv.org/abs/2210.14140
    - In practice this method produces pretty good results, just like **Min P** does.
  - **Temperature Last**:
    - We should probably emphasize that *Temperature Last* is almost always the sensible thing to do: pick the set of plausible tokens first, then tweak their relative probabilities (actually logits). Don't tweak the full distribution first, and then pick the token set from that, because this tends to amplify the probability of an incoherent response too much (which is what happens if *Temperature Last* is off).
  - **CFG**: Classifier Free Guidance.
    - Should also explain what it does... at least ooba uses CFG to control the strength of the negative prompt. Should check the source code for what ST does with it.
- *User Settings ⊳ Advanced*:
  - **No Text Shadows**:
    - Obvious, but missing a tooltip. **How should we treat such cases? Ignore them?**
  - **Visual Novel Mode**:
    - "Reduces chat height, and puts a static sprite behind the chat window."
    - ~**HALP!** What exactly does VN mode do, and how does it relate to group chats? What does it do in a 1-on-1 chat? This maybe needs a link to the manual, or something.~
  - **Expand Message Actions**:
    - "Message actions are context items for chat messages, usually hidden behind "...". Enable this to always show them."
    - ~**HALP!** What are message actions?~
  - **Zen Sliders**:
    - "Alternative UI for numeric sampling parameters with fewer steps."
  - **Mad Lab Mode**:
    - "Entirely unrestrict all numeric sampling parameters."
  - **Message Timer**:
    - "Time the AI's message generation, and show the duration in the chat log."
  - **Chat Timestamps**:
    - Obvious, but missing a tooltip.
  - **Model Icons**:
    - "Display an icon for the API that generated the message."
  - **Message IDs**:
    - "Show message numbers in the chat log."
  - **Message Token Count**:
    - "Show number of tokens in each message in the chat log."
  - **Compact Input Area**:
    - "Single-row message input area. Mobile view only, no effect on PC."
    - ~Nothing happens when toggling this on PC. It does have a mobile icon next to it, though - does that mean "mobile only"? What should this do?~
  - **Characters Hotswap**:
    - "In the Character Management panel, show quick selection buttons for favorited characters."
  - **Tags as Folders**:
    - ~**HALP!** What are tags? How to use them? Link to manual?~
    - Cohee explained this as: *"Assign tags to groups or characters from respective editing menus. Then use tag filters in the list."* Try it and figure out a tooltip.
  - **Message Sound**:
    - "Play a sound when a message generation finishes."
    - ~Has a link to the manual already. Could extract a one-line summary from there to use as a tooltip here.~
  - **Background Sound Only**:
    - "Only play a sound when ST's browser tab is unfocused."
  - **Custom CSS**:
    - "Apply a custom CSS style to all of the ST GUI."
    - ~What is the scope where the custom style applies? Just MovingUI, or the whole ST GUI?~
    - Where to get an example style to learn how to make new ones?
  - **Example Messages Behavior**:
    - Obvious, but missing a tooltip
  - **Advanced Character Search**:
    - "Use fuzzy matching, and search characters in the list by all data fields, not just by a name substring."
  - **Never resize avatars**:
    - "Avoid cropping and resizing imported character images. Default: crop/resize to 400x600."
  - **Show avatar filenames**:
    - "Show actual file names on the disk, in the characters list display only."
    - ~This seems to affect the *Character Management* panel only, not *Character Expressions* sprites?~
  - **Import Card Tags**:
    - "Prompt to import embedded card tags on character import. Otherwise embedded tags are ignored." 
    - ~Something to do with the PNG character card feature?~
  - **Spoiler Free Mode**:
    - "Hide character definitions from the editor panel behind a spoiler button."
  - **"Send" to Continue**:
    - Sending the message to the AI continues the last message instead of generating a new one? If this is enabled, how do you generate a new one?
    - According to Cohee, this feature was kept for backward compatibility since some users thought it important. Try it out and write a tooltip...
  - **Quick "Continue" button**:
    - "Show a button in the input area to ask the AI to continue (extend) its last message."
  - **Swipes**:
    - "Allow generating alternative responses from the AI before choosing which one to commit. Shows arrow buttons next to the AI's last message."
  - **Gestures**:
    - "Allow using swiping gestures on the last in-chat message to trigger swipe generation. Mobile only, no effect on PC."
    - Needs the mobile-only icon next to it!
  - **Auto-load Last Chat**:
    - Obvious, but missing a tooltip.
  - **Auto-scroll Chat**:
    - Obvious, but missing a tooltip.
  - **Auto-save Message Edits**:
    - "Saves edits to messages without confirmation as you type."
    - ~When does the autosave happen?~
  - **Confirm Message Deletion**:
    - Obvious, but missing a tooltip.
  - **Auto-fix Markdown** = ???
    - What exactly does it fix in Markdown, and using what algorithm?
  - **Render Formulas**:
    - "Render LaTeX and AsciiMath equation notation in chat messages. Powered by KaTeX."
  - **Show {{char}}: in responses**:
    - Obvious, but missing a tooltip.
  - **Show {{user}}: in responses**:
    - Obvious, but missing a tooltip.
  - **Show tags in responses**:
    - "Skip encoding < and > characters in message text, allowing a subset of HTML markup as well as Markdown."
  - **Log prompts to console**:
    - Obvious, but missing a tooltip.
  - **Auto-swipe**:
    - Obvious once you expand the panel and look at the available settings, but missing a tooltip.
    - "Automatically reject and re-generate AI message based on configurable criteria."
  - **Reload Chat**:
    - "Reload and redraw the currently open chat."
    - ~What exactly gets reloaded when you press this button?~

### Priority

Low (Nice-to-have)

### Is this something you would be keen to implement?

Yes!

**EDIT**: Updated list based on reply by @Cohee1207. Thanks!

**Comment by Cohee1207** [comment]
Looks like you want to conquer and master ALL of the UI and leave no stone unturned.

* AI Response Configuration - look them up in oobabooga UI.
* MUI Preset - Moving UI preset selector (predefined/saved draggable positions).
* Visual Novel Mode - reduces chat height and puts a static sprite behind the chat window.
* Expand Message Actions - message actions are context items for chat messages. Usually, they are hidden behind "..."
* Zen Sliders - alternative UI for numeric sampling parameters with fewer steps.
* Mad Lab Mode - entirely unrestricted numeric sampling parameters.
* Model Icons - displays an icon for the API that generated the message.
* Compact Input Area - single row message input area. For mobile view only.
* Tags as Folders - assign tags to groups or characters from respective editing menus. Then use tag filters in the list. 
* Message Sound - plays a sound when a message generation finishes.
* Background Sound Only - only play a sound when a browser tab is unfocused.
* Custom CSS - apply custom styles to all the ST UI.
* Advanced Character Search - uses fuzzy matching and searches characters in the list by all data fields, not just by a name substring.
* Never resize avatars - avoid cropping and resizing imported character images. Default: 400x600.
* Show avatar filenames - only characters list display, shows actual file names on the disk.
* Import Card Tags - prompts to import embedded card tags on character import. Ignores embedded tags otherwise.
* Spoiler Free Mode - hides character definitions from the editor panel behind a spoiler button.
* "Send" to Continue - this feature was once removed but was brought back once a couple of angry people yelled at me.
* Gestures - (mobile only) allow using swiping gestures on the last in-chat message to trigger swipe generation.
* Auto-save Message Edits - saves edits to messages without the confirmation as you type.
* Show tags in responses - doesn't encode `<` and `>` characters in the message texts, allowing a subset of HTML markup as well as Markdown.
* Reload Chat - reloads and redraws a currently open chat.

**Comment by Technologicat** [comment]
Just a perfectionist. UX is important.

I'll see what I can do. Some comments:

> Expand Message Actions 

Aaaah, I know what you mean now. I'd forgotten about that "..." menu.

> Mad Lab Mode - entirely unrestricted numeric sampling parameters.

Thanks, that was another wildly wrong first association for me. Extrapolating from the name, the only thing I could think of was [this](https://tvtropes.org/pmwiki/pmwiki.php/Main/MadLibsDialogue), which didn't make sense here.

> Background Sound Only - only play a sound when a browser tab is unfocused.

**EDIT**: Ah, so this didn't have anything to do with BGM or ambience.

> Spoiler Free Mode - hides character definitions from the editor panel behind a spoiler button.

Ah, that's an interesting option.

> "Send" to Continue - this feature was once removed but was brought back once a couple of angry people yelled at me.

Yeah, generally happens when removing any feature.

Thanks for the explanations!

**EDIT**: I'll update the list in the first post later.
**EDIT**: ...aaand done. Now to find some time to implement these. And of course, in the meantime, more information is always welcome!

**Comment by Technologicat** [comment]
One more note:

1. IIRC, a couple of months ago ooba used to delegate the sampler parameter explanations to the Transformers documentation, but seems they've stepped up their game recently. The new link in the webui is [to ooba's own wiki](https://github.com/oobabooga/text-generation-webui/wiki/03-%E2%80%90-Parameters-Tab), which has a decent explanation of many of the parameters. I'll synthesize something from that and from what I've read in the relevant papers and blog posts.
2. That's all for new bug reports from me for now. :)

**Comment by Technologicat** [comment]
@Cohee1207: I'm preparing a PR for the first batch of new tooltips.

I'm also trying to wrap my head around the rather puzzling shape of some things here. So here's some software archaeology, please correct me if I'm wrong:

- **Reasonable uninformed prior**: Different modes for achieving the same thing behave similarly.
- **Puzzling evidence**: In *Summarize*, some Main API features are missing when using the Extras API:
  - "Skip WI/AN" does nothing. The checkbox remains visible, but the setting isn't even consulted when preparing the Extras API call.
  - "Summarize now" is missing, and is actually hidden in Extras mode.
- **Posterior hypothesis**:
  - In the ancient past (early 2023-ish?), summaries were produced by the dedicated `summarize` Extras module, which used a dedicated AI checkpoint specialized to this task.
  - This season's approach is to repurpose the main LLM to create the summary by instructing it appropriately.

From which I conclude that the Extras API option has not been kept up to feature parity, and it has only been preserved for backward compatibility.

If so, perhaps this should be mentioned in the UI.

**Comment by Cohee1207** [comment]
> "Summarize now" is missing, and is actually hidden in Extras mode.

By design. Extras way to handle summaries is entirely different. Main API uses a single LLM prompt, and Extras API is double buffered with the previous summary being included as a part of the next request. That is because BART (model used for summarization) has only 1024 tokens of context. Same with "Skip WI/AN" - it is not included in Extras API requests.

**Comment by Technologicat** [comment]
Thanks. If WI/AN are always omitted in Extras mode, I suppose we could hide the checkbox in Extras mode?

**Comment by Technologicat** [comment]
Resolved by #1726.

There's always more, though... currently tracked in #1739.
